runners/direct-java/src/main/java/org/apache/beam/runners/direct/DirectRunner.java

```diff
@@ -364,5 +366,17 @@ public State awaitCompletion() throws Throwable {
       }
       return state;
     }
+
+    @Override
+    public State cancel() throws IOException {
+      throw new UnsupportedOperationException("DirectPipelineResult does not support cancel.");
```

Comment:
ditto as comments below

DirectRunner.run() can be non blocking. And, it should be able to force shutting down itself.